src/traverseDelegateProperties.js

```diff
@@ -0,0 +1,194 @@
+  return pathStrSegment.substr(
+    0,
+    // the name goes up to the start of the array bracket: 'someArrayName[]'
+    pathStrSegment.indexOf('[')
+  );
+}
+
+function isObject(value) {
+  if (value == null) {
+    return false;
+  }
+
+  return Boolean(
+    typeof value === 'object' &&
+      Object.prototype.toString.call(value) === '[object Object]' &&
+      !Array.isArray(value)
+  );
+}
+
+function isString(value) {
+  var type = typeof value;
+  return Boolean(
+    type === 'string' ||
+      (type === 'object' &&
+        value != null &&
+        !Array.isArray(value) &&
+        Object.prototype.toString.call(value) === '[object String]')
+  );
+}
+
+function traverseDelegateProperties(pathArray, settingsSlice) {
+  // A string of the form a.b[4].c.names[2] is split into
+  // [a, b[4], c, names, [2]]
+  if (!Array.isArray(pathArray) || !pathArray.length) {
+    return undefined;
+  }
+
+  var nextPathKey = pathArray[0];
+  var nextPropertyValue;
+
+  if (isArrayReference(nextPathKey)) {
+    var nextPathKeyParts = sanitizeArrayKey(nextPathKey);
+    nextPropertyValue =
+      settingsSlice[nextPathKeyParts.arrName][nextPathKeyParts.arrIndex];
+  } else {
+    // objects
+    nextPropertyValue = settingsSlice[nextPathKey];
+  }
+
+  if (pathArray.length === 1) {
+    return nextPropertyValue;
+  }
+
+  return traverseDelegateProperties(pathArray.slice(1), nextPropertyValue);
+}
+
+function pluckSettingsValue(pathString, settings) {
+  if (!isString(pathString) || !isObject(settings)) {
+    return undefined;
+  }
+
+  return traverseDelegateProperties(pathString.split('.'), settings);
+}
+
+// function pushValueIntoSettings(pathString, settings, value) {
+//   if (!isString(pathString) || !isObject(settings)) {
+//     return undefined;
+//   }
+//
+//   var settingsCopy = objectAssign({}, settings);
+//
+//   var pathSegments = pathString.split('.');
+//   var nextSettingsLevel = settingsCopy; // 2 references now point to "settingsCopy"
+//
+//   // Pull off because this is the ultimate variable that needs to be granted the passed in value
+//   var finalSegment = pathSegments.pop();
+//
+//   // traverse down to the inner-most path
+//   pathSegments.forEach(function (pathSegment, index) {
+//     if (isArrayReference(pathSegment)) {
+//       var nextPathKeyParts = sanitizeArrayKeyAndSplit(pathSegment);
+//       nextSettingsLevel =
+//         nextSettingsLevel[nextPathKeyParts.arrName][nextPathKeyParts.arrIndex];
+//     } else {
+//       nextSettingsLevel = nextSettingsLevel[pathSegment];
+//     }
+//   });
+//
+//   // update the value
+//   if (isArrayReference(finalSegment)) {
+//     var nextPathKeyParts = sanitizeArrayKeyAndSplit(finalSegment);
+//     nextSettingsLevel[nextPathKeyParts.arrName][
+//       nextPathKeyParts.arrIndex
+//     ] = value;
+//   } else {
+//     nextSettingsLevel[finalSegment] = value;
+//   }
+//
+//   return settingsCopy;
+// }
+
+function pushValueIntoSettings(
```

Comment:
This method I think has a few problems:
* If the value that needs to be changed is not a string, it still calls `decorateWithDynamicHost` which I think will result in an error. In the test the error is not visible because decorateWithDynamicHost is mocked to return a string no matter what parameters are used. I think in case the filePath gets a result that is not a string, we should not touch the settings.
* The second problem is that if you have a settings like `{a: { b: 'b1'}, b: 'b2'}` and you call the function with `a.b`, the end result will be wrong. You will get back an object that has the `b` top replaced.

I was thinking to do something more close to this:
```
function pushValueIntoSettings(
  pathSegments,
  settings,
  decorateWithDynamicHost
) {
  if (pathSegments.length === 0) {
    return isString(settings) ? decorateWithDynamicHost(settings) : settings;
  }

  var currentKey = pathSegments.shift();
  var isKeyAnArray = false;
  if (isArrayReference(currentKey)) {
    currentKey = sanitizeArrayKey(currentKey);
    isKeyAnArray = true;
  }

  if (
    !settings.hasOwnProperty(currentKey) ||
    (isKeyAnArray && !Array.isArray(settings[currentKey]))
  ) {
    return settings;
  }

  if (isKeyAnArray) {
    settings[currentKey] = settings[currentKey].map(function (
      arrayEntryObject
    ) {
      return pushValueIntoSettings(
        pathSegments.slice(),
        arrayEntryObject,
        decorateWithDynamicHost
      );
    });
  } else {
    settings[currentKey] = pushValueIntoSettings(
      pathSegments.slice(),
      settings[currentKey],
      decorateWithDynamicHost
    );
  }

  return settings;
}

module.exports = {
  pluckSettingsValue: pluckSettingsValue,
  pushValueIntoSettings: function (
    filePathString,
    settings,
    decorateWithDynamicHost
  ) {
    window.decorateWithDynamicHost = decorateWithDynamicHost;
    if (
      !isString(filePathString) ||
      !filePathString.length ||
      !isObject(settings)
    ) {
      return settings;
    }

    var pathSegments = filePathString.split('.');

    return pushValueIntoSettings(
      pathSegments,
      settings,
      decorateWithDynamicHost
    );
  }
};
```

Reply:
I guess the biggest question becomes how much do we trust Forge, and do we really want to try masking problems? If Forge is telling me there's a path that needs to be made into an absolute URL, I'm implicitly trusting that it knows what it's telling me is accurate.

Reply:
Forge doesn't do any validation. I spoke with Spencer this morning and he said:
`Forge is expected to leave the object alone if the path is not expected. Turbine should have a test for this and do the same.`

Reply:
The second problem I listed is not something Forge can fix.

Reply:
Yea, I've got some code here to resolve that. Let me reply to decorate about point 1.